File: app/src/main/java/com/kimjisub/launchpad/activity/LaunchpadActivity.java
package com.kimjisub.launchpad.activity;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.hardware.usb.UsbConstants;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbEndpoint;
import android.hardware.usb.UsbInterface;
import android.hardware.usb.UsbManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.kimjisub.launchpad.R;
import com.kimjisub.launchpad.databinding.ActivityUsbmidiBinding;
import com.kimjisub.launchpad.manager.PreferenceManager;
import com.kimjisub.launchpad.midi.MidiConnection;
import com.kimjisub.launchpad.midi.controller.MidiController;
import com.kimjisub.launchpad.midi.driver.DriverRef;
import com.kimjisub.launchpad.midi.driver.LaunchpadMK2;
import com.kimjisub.launchpad.midi.driver.LaunchpadPRO;
import com.kimjisub.launchpad.midi.driver.LaunchpadS;
import com.kimjisub.launchpad.midi.driver.MasterKeyboard;
import com.kimjisub.launchpad.midi.driver.MidiFighter;
import com.kimjisub.launchpad.midi.driver.Noting;
import com.kimjisub.manager.Log;

import java.util.Iterator;
import java.util.Objects;

public class LaunchpadActivity extends BaseActivity {

	ActivityUsbmidiBinding b;
	LinearLayout[] LL_Launchpad;
	LinearLayout[] LL_mode;

	@SuppressLint({"CutPasteId", "StaticFieldLeak"})
	void initVar() {
		LL_Launchpad = new LinearLayout[]{
				findViewById(R.id.s),
				findViewById(R.id.mk2),
				findViewById(R.id.pro),
				findViewById(R.id.midifighter),
				findViewById(R.id.piano)
		};
		LL_mode = new LinearLayout[]{
				findViewById(R.id.speedFirst),
				findViewById(R.id.avoidAfterimage)
		};
	}

	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		b = setContentViewBind(R.layout.activity_usbmidi);
		initVar();

		MidiConnection.setListener(new MidiConnection.Listener() {
			@Override
			public void onConnectedListener() {
				b.err.setVisibility(View.GONE);
			}

			@Override
			public void onChangeDriver(Class cls) {
				selectDriver(cls);
			}

			@Override
			public void onChangeMode(int mode) {
				selectMode(mode);
			}

			@Override
			public void onUiLog(String log) {
				b.info.append(log + "\n");
			}
		});

		MidiConnection.setMode(PreferenceManager.LaunchpadConnectMethod.load(LaunchpadActivity.this));

		Intent intent = getIntent();
		MidiConnection.initConnection(intent, (UsbManager) getSystemService(Context.USB_SERVICE));


		(new Handler()).postDelayed(this::finish, 2000);
	}

	// Select Driver /////////////////////////////////////////////////////////////////////////////////////////

	public void selectDriver(View v) {
		int index = Integer.parseInt((String) v.getTag());
		selectDriver(new Class[]{LaunchpadS.class, LaunchpadMK2.class, LaunchpadPRO.class, MidiFighter.class, MasterKeyboard.class}[index]);
	}


	public void selectDriver(Class cls) {
		int index = 0;
		switch (cls.getSimpleName()) {
			case "LaunchpadS":
				index = 0;
				break;
			case "LaunchpadMK2":
				index = 1;
				break;
			case "LaunchpadPRO":
				index = 2;
				break;
			case "MidiFighter":
				index = 3;
				break;
			case "MasterKeyboard":
				index = 4;
				break;
		}

		for (int i = 0; i < LL_Launchpad.length; i++) {
			if (index == i) {
				LL_Launchpad[i].setBackgroundColor(color(R.color.gray1));
				int count = LL_Launchpad[i].getChildCount();
				for (int j = 0; j < count; j++) {
					TextView textView = (TextView) LL_Launchpad[i].getChildAt(j);
					textView.setTextColor(color(R.color.background1));
				}
			} else {
				LL_Launchpad[i].setBackgroundColor(color(R.color.background1));
				int count = LL_Launchpad[i].getChildCount();
				for (int j = 0; j < count; j++) {
					TextView textView = (TextView) LL_Launchpad[i].getChildAt(j);
					textView.setTextColor(color(R.color.gray1));
				}
			}
		}
	}


	// Select Mode /////////////////////////////////////////////////////////////////////////////////////////

	public void selectModeXml(View v) {
		selectMode(Integer.parseInt((String) v.getTag()));
	}

	public void selectMode(int mode) {
		for (int i = 0; i < LL_mode.length; i++) {
			if (mode == i) {
				LL_mode[i].setBackgroundColor(color(R.color.gray1));
				int count = LL_mode[i].getChildCount();
				for (int j = 0; j < count; j++) {
					TextView textView = (TextView) LL_mode[i].getChildAt(j);
					textView.setTextColor(color(R.color.background1));
				}
			} else {
				LL_mode[i].setBackgroundColor(color(R.color.background1));
				int count = LL_mode[i].getChildCount();
				for (int j = 0; j < count; j++) {
					TextView textView = (TextView) LL_mode[i].getChildAt(j);
					textView.setTextColor(color(R.color.gray1));
				}
			}
		}

		PreferenceManager.LaunchpadConnectMethod.save(LaunchpadActivity.this, mode);
	}



	// Controller /////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public void onResume() {
		super.onResume();
		initVar();
	}

	@SuppressLint("StaticFieldLeak")
	@Override
	public void onDestroy() {
		super.onDestroy();
	}
}

File: app/src/main/java/com/kimjisub/launchpad/activity/LaunchpadActivity.kt
package com.kimjisub.launchpad.activity

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.hardware.usb.UsbManager
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.kimjisub.launchpad.R.color
import com.kimjisub.launchpad.R.layout
import com.kimjisub.launchpad.manager.PreferenceManager.LaunchpadConnectMethod
import com.kimjisub.launchpad.midi.MidiConnection.Listener
import com.kimjisub.launchpad.midi.MidiConnection.initConnection
import com.kimjisub.launchpad.midi.MidiConnection.setListener
import com.kimjisub.launchpad.midi.MidiConnection.setMode
import com.kimjisub.launchpad.midi.driver.*
import kotlinx.android.synthetic.main.activity_usbmidi.*

class LaunchpadActivity : BaseActivity() {
	private val LL_Launchpad: Array<LinearLayout> by lazy {
		arrayOf(
			BTN_S,
			BTN_Mk2,
			BTN_Pro,
			BTN_Midifighter,
			BTN_Piano
		)
	}

	private val LL_mode: Array<LinearLayout> by lazy {
		arrayOf(
			BTN_speedFirst,
			BTN_avoidAfterimage
		)
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(layout.activity_usbmidi)
		setListener(object : Listener {
			override fun onConnectedListener() {
				RL_err.visibility = View.GONE
			}

			override fun onChangeDriver(cls: Class<*>) {
				selectDriver(cls)
			}

			override fun onChangeMode(mode: Int) {
				selectMode(mode)
			}

			override fun onUiLog(log: String) {
				TV_log.append(log + "\n")
			}
		})
		setMode(LaunchpadConnectMethod.load(this@LaunchpadActivity))
		val intent: Intent? = intent
		initConnection(intent!!, (getSystemService(Context.USB_SERVICE) as UsbManager))
		Handler().postDelayed({ finish() }, 2000)
	}

	// Select Driver /////////////////////////////////////////////////////////////////////////////////////////


	fun selectDriver(v: View) {
		val index = Integer.parseInt(v.tag as String)
		selectDriver(
			arrayOf<Class<*>>(
				LaunchpadS::class.java,
				LaunchpadMK2::class.java,
				LaunchpadPRO::class.java,
				MidiFighter::class.java,
				MasterKeyboard::class.java
			)[index]
		)
	}

	fun selectDriver(cls: Class<*>) {
		var index = 0
		when (cls.simpleName) {
			"LaunchpadS" -> index = 0
			"LaunchpadMK2" -> index = 1
			"LaunchpadPRO" -> index = 2
			"MidiFighter" -> index = 3
			"MasterKeyboard" -> index = 4
		}
		for (i in LL_Launchpad.indices) {
			if (index == i)
				changeViewColor(LL_Launchpad[i], color.gray1, color.background1)
			else
				changeViewColor(LL_Launchpad[i], color.background1, color.gray1)

		}
	}


	// Select Mode /////////////////////////////////////////////////////////////////////////////////////////


	fun selectModeXml(v: View) {
		selectMode(Integer.parseInt(v.tag as String))
	}

	fun selectMode(mode: Int) {
		for (i in LL_mode.indices) {
			if (mode == i)
				changeViewColor(LL_mode[i], color.gray1, color.background1)
			else
				changeViewColor(LL_mode[i], color.background1, color.gray1)
		}
		LaunchpadConnectMethod.save(this@LaunchpadActivity, mode)
	}

	// Functions /////////////////////////////////////////////////////////////////////////////////////////

	private fun changeViewColor(layout: ViewGroup, backgroundColorRes: Int, textColorRes: Int) {
		layout.setBackgroundColor(colors.get(backgroundColorRes))
		val count = layout.childCount
		for (j in 0 until count) {
			val textView = layout.getChildAt(j) as TextView
			textView.setTextColor(colors.get(textColorRes))
		}
	}

	// ActivityCycle /////////////////////////////////////////////////////////////////////////////////////////

	@SuppressLint("StaticFieldLeak")
	override fun onDestroy() {
		super.onDestroy()
	}
}